content/posts/lets-code-contact-form-in-phoenix-part-two.md

```diff
@@ -0,0 +1,240 @@
+  # to an empty map.
+  @spec change_message(map) :: Ecto.Changeset.t()
+  def change_message(fields \\ %{}) do
+    %Message{}
+    |> Message.changeset(fields)
+  end
+
+  # This function was updated to use the new `change_message` function.
+  def send_message(fields) when is_map(fields) do
+    fields
+    |> change_message()
+    |> Ecto.Changeset.apply_action(:insert)
+  end
+end
+```
+
+Re-running the tests should result in success. Let's take a step back now and review the final product.
+
+## The Final Product
+
+```elixir
+# lib/my_app/support.ex
+
+defmodule LearnU.Support do
+  @moduledoc """
+  The Support module is the interface to customer support for the application.
+  Look to this module when you want to provide support to application users.
+  """
+
+  alias __MODULE__.Message
+
+  @doc """
+  Returns a changeset for a Support Message.
+  """
+  @spec change_message(map) :: Ecto.Changeset.t()
+  def change_message(fields \\ %{}) do
+    %Message{}
+    |> Message.changeset(fields)
+  end
+
+  @doc """
+  Sends a message with given fields to the support team. Fields should include:
+  - `email`: Email for the user submitting the message for the support team to respond.
+  - `subject`: A short, general title as to the reason for support
+  - `body`: A longer description as to the reason for support and how the
+            support team might be able to help.
+
+  If the message is invalid, an `{:error, changeset}` is returned. Returns
+  `{:ok, message}` on success.
+  """
+  @spec send_message(map) :: {:ok, Message.t()} | {:error, Ecto.Changeset.t()}
+  def send_message(fields) when is_map(fields) do
+    fields
+    |> change_message()
+    |> Ecto.Changeset.apply_action(:insert)
+  end
+end
+```
+
+```elixir
+# test/my_app/support_test.exs
+
+defmodule LearnU.SupportTest do
+  use LearnU.DataCase, async: true
+
+  alias LearnU.Support
+
+  test "changing a message" do
+    assert %Ecto.Changeset{data: %Support.Message{}} = Support.change_message()
+  end
+
+  describe "sending a message" do
+    test "with valid fields" do
+      fields = %{
+        email: "barry@bluejeans.test",
+        subject: "Halp Me",
+        body: "I am having problems."
+      }
+
+      assert {:ok, %Support.Message{} = message} = Support.send_message(fields)
+      assert message.email == "barry@bluejeans.test"
+      assert message.subject == "Halp Me"
+      assert message.body == "I am having problems."
+    end
+
+    test "with invalid fields" do
+      assert {:error, cset} = Support.send_message(%{})
+
+      refute cset.valid?
+      assert "can't be blank" in errors_on(cset).email
+    end
+  end
+end
+```
+
+This post covered the foundation of the public interface for the `Support` domain of the Elixir application. I created the functions to change and send a message, but have not implemented the actual sending of an email to the support team. I will cover this in the next installment of this Let's Code series. Thanks for joining me! 👋
```

Comment:
While reading, I started wondering about the fact we weren't doing any actual sending a bit earlier. It could be work calling that out earlier as well (pre-refactoring maybe?), but I'm not sure it matters too much since you do call it out.

Reply:
Yea, I was afraid this post was going to end up too short without the email sending bit, but the email sending requires quite a bit of setup. When I saw this post come out around 5-7 minutes, I felt like that was a good enough length to cut it there. I can focus the next post on the setup required for sending email using the Bamboo library. I'll toy around with calling that out a bit sooner.